stage0/src/acpi.rs

```diff
@@ -21,21 +21,106 @@ use crate::fw_cfg::FwCfg;
-// structure, and where it should be placed.
-const RSDP_SIZE: usize = 36;
-// Rest of the EBDA. The full size of EBDA is 128 KiB, but let's reserve the whole 1 KiB for the
-// RSDP.
+use zerocopy::{AsBytes, FromBytes};
+
+/// ACPI Root System Description Pointer.
+///
+/// Used to locate either the RSDT or XSDT in memory.
+///
+/// See Section 5.2.5 in the ACPI specification, Version 6.5 for more details.
+#[derive(FromBytes, AsBytes)]
+#[repr(C, packed)]
+pub struct Rsdp {
+    /// Signature: "RSD PTR " (note the trailing space).
+    signature: [u8; 8],
+
+    /// Checksum for fields defined in the ACPI 1.0 specification.
+    checksum: u8,
+
+    /// OEM-supplied identification string.
+    oemid: [u8; 6],
+
+    /// Revision of this structure.
+    ///
+    /// ACPI 1.0 value is zero, ACPI 2.0 value is 2.
+    revision: u8,
+
+    /// 32-bit physical address of the RSDT.
+    rsdt_address: u32,
+
+    // ACPI 2.0 fields.
+    /// Length of the table, including the header.
+    length: u32,
+
+    /// 64-bit physical address of the XSDT.
+    xsdt_address: u64,
+
+    /// Checksum of the entire table, including both checksum fields.
+    extended_checksum: u8,
+
+    /// Reserved
+    _reserved: [u8; 3],
+}
+static_assertions::assert_eq_size!(Rsdp, [u8; 36usize]);
+
+impl Rsdp {
+    fn validate(&self) -> Result<(), &'static str> {
+        if &self.signature != b"RSD PTR " {
+            return Err("Invalid RSDP signature");
+        }
+        let len = if self.revision > 2 { self.length } else { 20 } as usize;
+
+        if len > size_of::<Rsdp>() {
+            return Err("invalid RSDP size");
+        }
+
+        let checksum = self.as_bytes()[..20]
+            .iter()
+            .fold(0u8, |lhs, &rhs| lhs.wrapping_add(rhs));
+
+        if checksum != 0 {
+            return Err("Invalid RSDP checksum");
+        }
+
+        if self.revision > 2 {
+            let checksum = self
+                .as_bytes()
+                .iter()
+                .fold(0u8, |lhs, &rhs| lhs.wrapping_add(rhs));
+
+            if checksum != 0 {
+                return Err("Invalid RSDP extended checksum");
+            }
+        }
+
+        // Check the pointer addresses; if they are valid, they should point within the EBDA.
+        // Safety: we will never dereference the pointer, we just need to know where it points to.
+        let ebda_base = unsafe { EBDA.as_ptr() } as usize;
+        if self.rsdt_address > 0
+            && ((self.rsdt_address as usize) < ebda_base
+                || (self.rsdt_address as usize) >= ebda_base + EBDA_SIZE)
+        {
+            return Err("Invalid RSDT address: does not point to within EBDA");
+        }
+        if self.xsdt_address > 0
+            && ((self.xsdt_address as usize) < ebda_base
+                || (self.xsdt_address as usize) >= ebda_base + EBDA_SIZE)
+        {
+            return Err("Invalid XSDT address: does not point to within EBDA");
+        }
```

Comment:
nit, might be to extract this into a private function but I don't feels strongly about it. :)